### PR TITLE
Reduce Dictionary's field size by 20%

### LINKED ISSUE #6750: Optimization discussion: Shrinking Dictionary by 3 fields
Right now [Dictionary](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/Dictionary.cs) has quite a few fields in its layout:

```cs
public class Dictionary<TKey, TValue>
{
    int[] buckets; // 1 word
    Entry[] entries; // 1 word
    int count; // 4 bytes
    int version; // 4 bytes
    int freeList; // 4 bytes
    int freeCount; // 4 bytes
    IEqualityComparer<TKey> comparer; // 1 word
    KeyCollection keys; // 1 word
    ValueCollection values; // 1 word
    object _syncRoot; // 1 word
}
```

This totals to 48 + 16 == 64 bytes for x64, for the fields alone. In fact [in my test](https://dotnetfiddle.net/vDDU0p) I'm finding the entire object (including method table and all) to total to 80 bytes, and that doesn't even include any of the arrays/other objects that are being allocated.

I'm experimenting [**here**](https://github.com/dotnet/corefx/compare/master...jamesqo:patch-2) with cutting out those last 4 fields and replacing them with a single field, `object comparerOrCollections`. Basically how it works is, if `Keys` / `Values` / `SyncRoot` has not been called yet, it will be set to the equality comparer. Otherwise, if `Keys` / `Values` are called (which any code that cares about perf [really shouldn't](https://github.com/dotnet/corefx/pull/7765#discussion_r59931593)), it's going to create a new, intermediate object that is responsible for lazy-initialization of `Keys` / `Values`. Since this intermediate object is never going to be seen by the outside world, we can improvise further and use it as the `SyncRoot`.

Obviously this will incur a bit more overhead for `Keys` / `Values`, as there will be an additional allocation + `CompareExchange` on lazy initialization, not to mention another pointer indirection. This could also affect usage of `comparer` since we don't know exactly which field it's in until we typecast.

The payoff for all of this is a 33% reduction in size of Dictionary (80 bytes -> 56).

Thoughts? @mikedn, @omariom, @benaadams? (Note the commit I linked to is basically a draft, I have not run tests with it yet since I don't want to put too much work in if this is not the way to go.)

### PR DESCRIPTION
**Pull request by jamesqo**
`Keys` / `Values`  properties of Dictionary should not be called by someone who cares about performance, since this is an additional allocation just to make it conform to an interface. Any code that accepts an `IEnumerable<TKey>` or `IEnumerable<TValue>` can be rewritten (albeit with less genericity) to accept an `IEnumerable<KeyValuePair<TKey, TValue>>` instead, and only use the Key/Value property.

Based on this assumption, I've aggregated the cached `Keys` / `Values` properties  of Dictionary into a new private class, `KeysAndValuesHolder`. This saves us 1 word-sized field. Since this class is never going to be viewed by the outside world, we can go a step further and use it as the `SyncRoot` which eliminates yet another field.

This takes the size of Dictionary (including method table and all) from 80 bytes => 64 bytes in memory for 64-bit systems. ([proof](https://dotnetfiddle.net/vDDU0p))

Drawbacks: If someone does use `Keys` / `Values`, then there will be an additional allocation + Interlocked.CompareExchange for the first-time initialization. Also, there will be a method call to `KeysAndValuesHolder.get_Keys` instead of a (previously inlined) field access, not to mention an extra pointer indirection.

Fixes #6750

Going to add extra tests to corefx for this soon.

cc @jkotas @mikedn @omariom @GSPP @benaadams @justinvp

### CODE REVIEW

**Comment by benaadams** [comment]
cc @stephentoub

File: src/mscorlib/src/System/Collections/Generic/Dictionary.cs [hunk 1]
```diff
@@ -815,6 +793,32 @@ public struct Enumerator: IEnumerator<KeyValuePair<TKey,TValue>>,
...
+            private ValueCollection values;
+
+            public KeysAndValuesHolder(Dictionary<TKey, TValue> dictionary)
+            {
+                Contract.Assert(dictionary != null);
+                this.dictionary = dictionary;
+            }
+
+            public KeyCollection Keys =>
+                keys ?? (keys = new KeyCollection(dictionary));
```
Comment (jamesqo) [inline src/mscorlib/src/System/Collections/Generic/Dictionary.cs]: Unfortunately, applying `AggressiveInlining` here seems to kill the inline of get_KeysAndValues. PerfView says the reason is "noinline per IL/cached result". Maybe I should apply it to both, but something is telling me I should trust the JIT compiler...

**Comment by stephentoub** [comment]
Thanks.  Though I'm not convinced this is a worthwhile change.  An empty Dictionary is 80 bytes on 64-bit, so at best this is saving 10%.  The moment you add a single item to a ```Dictionary<int, int>``` that size jumps to 192 bytes, at which point this is only a 4% saving, and the value of this continues to drop as the dictionary expands (e.g. adding 16 int,int pairs brings it up to 480 bytes, such that this is only a 1.5% memory saving).  And while even a small decrease in memory saving is good, there is a lot of code that uses Dictionary.Keys/Values.... just look at http://referencesource.microsoft.com/#mscorlib/system/collections/generic/dictionary.cs,9a3c0cb5c149c9f7,references , where within the framework there are almost 300 call sites to it, and such a change would make all of those more expensive.  Without more justification that shows this is really worth the tradeoff,I don't believe it's a change that should be made.

File: src/mscorlib/src/System/Collections/Generic/Dictionary.cs (hunk 1 above)
Comment (jamesqo) [inline src/mscorlib/src/System/Collections/Generic/Dictionary.cs]: What's even more bizarre, doing this:

```cs
public KeyCollection Keys
{
    [MethodImpl(AggressiveInlining)]
    get
    {
        return keys ?? InitializeKeys();
    }
}

[MethodImpl(NoInlining)]
private KeyCollection InitializeKeys()
{
    keys = new KeyCollection(dictionary);
    return keys;
}
```

seems to kill inlining of *both* `KeysAndValuesHolder.get_Keys` and `Dictionary.get_KeysAndValues`. No idea why.

File: src/mscorlib/src/System/Collections/Generic/Dictionary.cs [hunk 2]
```diff
@@ -815,6 +793,32 @@ public struct Enumerator: IEnumerator<KeyValuePair<TKey,TValue>>,
...
+        // This is separated into a new class to save on memory by reducing
+        // fields if Keys/Values are never called.
+        // Since it shouldn't be visible to the outside world, it also serves
+        // as the SyncRoot (also rarely used) which saves an additional field.
+        // The downside of this is that Keys/Values are called, then there will
+        // be an additional object allocation the first time, plus this class
+        // needs to be initialized with Interlocked.CompareExchange.
+        private sealed class KeysAndValuesHolder
+        {
+            private readonly Dictionary<TKey, TValue> dictionary;
```
Comment (mikedn) [inline src/mscorlib/src/System/Collections/Generic/Dictionary.cs]: I don't think this field is needed. `Keys` and `Values` could be methods talking a dictionary parameter.

**Comment by mikedn** [comment]
> Keys / Values properties of Dictionary should not be called by someone who cares about performance, since this is an additional allocation just to make it conform to an interface

I told you before that this isn't true.

**Comment by jamesqo** [comment]
@mikedn You said

> Anytime you need an IEnumerable of keys or values you can just use those collections directly without creating an intermediary list or array.

I responded to that in my post, you can change such code to accept a `IEnumerable<KeyValuePair<TKey, Value>>` instead (which Dictionary implements) and only use the Key/Value property.

**Comment by mikedn** [comment]
> I responded to that in my post, you can change such code to accept a IEnumerable<KeyValuePair<TKey, Value>> instead (which Dictionary implements) and only use the Key/Value property.

No I can't. The code may not be written by me or is supposed to accept an `IEnumerable<T>` or `ICollection<T>` because it may be called with a list or who knows what other collection type.

**Comment by stephentoub** [comment]
> you can change such code to accept a IEnumerable<KeyValuePair<TKey, Value>> instead (which Dictionary implements) and only use the Key/Value property.

Not if you're passing it to something that expects an ```IEnumerable<TKey>``` or ```IEnumerable<TValue>```.

But even if that were the case, this argument is based on expecting all existing usage to change.

**Comment by jamesqo** [comment]
I'm deciding to close this after the arguments presented here; although it's true all code can't just be rewritten to accept an IEnumerable of `KeyValuePair`, I don't think this PR really degrades performance that much for `get_Keys` or `get_Values` (assuming I would be able to get `KeysAndValuesHolder.get_Keys` to somehow inline). However, as @stephentoub mentioned this will have a small benefit as soon as some entries are added, and a typical usage of this class is to have `string` as the TKey so the benefit after, say, 16 entries will be even less than 3%. The only scenario in which this will pay off significantly (> 10%) is when someone allocates a bunch of dictionaries and never adds anything to them, which is kind of pointless. Based on all of this, I'm closing this PR.

**Comment by GSPP** [comment]
@jamesqo  It was a good attempt. If we were creating the class now we could consider creating a fresh value each time even. It would be no worse than a `EnumerateKeys()` method. But that's too late.

**Pull request closed without merging**